model/src/main/kotlin/PackageReference.kt

```diff
@@ -72,16 +73,28 @@ data class PackageReference(
     fun collectDependencies(
         maxDepth: Int = -1,
         filterPredicate: (PackageReference) -> Boolean = { true }
-    ): SortedSet<Identifier> =
-        dependencies.fold(sortedSetOf()) { refs, ref ->
-            refs.also {
-                if (maxDepth != 0) {
-                    if (filterPredicate(ref)) it += ref.id
-                    it += ref.collectDependencies(maxDepth - 1, filterPredicate)
-                }
+    ): Set<Identifier> {
+        val result = mutableSetOf<Identifier>()
+
+        val queue = Stack<Pair<PackageReference, Int>>()
+        queue.push(Pair(this, 0))
+
+        while (queue.isNotEmpty()) {
+            val (pkg, level) = queue.pop()
+
+            if (filterPredicate(pkg)) {
+                result.add(pkg.id)
+            }
+
+            val dependenciesLevel = level + 1
+            if (maxDepth < 0 || dependenciesLevel <= maxDepth) {
+                queue.addAll(pkg.dependencies.map { Pair(it, dependenciesLevel) })
```

Comment:
This stores the `dependenciesLevel` for each dependency of the same level in the `Pair`. I wonder if we could be slightly more memory efficient and store the `dependenciesLevel` only once for all dependencies of the same level (in a dependency branch) by doing something like:

    fun collectDependencies(
        maxDepth: Int = -1,
        filterPredicate: (PackageReference) -> Boolean = { true }
    ): Set<Identifier> {
        val queue = Stack<Pair<SortedSet<PackageReference>, Int>>()
        queue.push(Pair(dependencies, 0))

        val result = mutableSetOf<Identifier>()
        while (queue.isNotEmpty()) {
            val (depsOnLevel, level) = queue.pop()

            result += depsOnLevel.filter(filterPredicate).map { it.id }

            val dependenciesLevel = level + 1
            if (maxDepth < 0 || dependenciesLevel <= maxDepth) {
                queue.addAll(depsOnLevel.map { it.dependencies to dependenciesLevel })
            }
        }

        return result
    }

Reply:
I prefer to keep the code `as-is` because I belief the potential memory gain you refer to is neglectible and because in my point of view the code `as-is` is a bit simpler and thus easier to grasp.

Reply:
I agree that the memory gain probably is rather small, but I have to admit I had problems understanding the code, and why you chose a stack as a data structure in particular. What I found hard to see through is the fact that you `pop` only one item per `while` iteration, but potentially add multiple items per iteration. When looking at the stack for each iteration, this leads to an interleaved mix of dependency levels on the stack, which is a bit odd to comprehend IMO.

Reply:
Since this is about performance you should replace this line
```kotlin
queue.addAll(pkg.dependencies.map { Pair(it, dependenciesLevel) })
```

with this line to avoid the creation of temporary lists:
```kotlin
pkg.dependencies.forEach { queue.add(Pair(it, dependenciesLevel) } // or push instead of add
```

Reply:
Regarding `enqueing multiple items per `while` iteration is how all the BFS algrithms are written and I do not see what the issue is with that. I do not see even an alternative without making the algorithm `recursive` again. It would be cool if you could tell me what alternative there is available.

Regarding the `Stack` I agree that picking `LIFO` rather than `FIFO` is not so nice, let me change it to `FIFO` to align with BFS and to have the guarantees that (1) the level in the queue are sorted acsending and (2) the min and max enqueue level differs at most by one.

Reply:
Thanks, I like the new `FIFO` code much better.

Reply:
Same here.